projects/capnproto/fuzz-run.c++

```diff
@@ -0,0 +1,71 @@
+#include "test-util.h"
+#include <kj/main.h>
+#include "serialize.h"
+#include <capnp/test.capnp.h>
+#include <unistd.h>
+
+namespace capnp {
+    namespace _ {
+        namespace {
+
+  class ParseErrorCatcher: public kj::ExceptionCallback {
+  public:
+    ParseErrorCatcher(kj::ProcessContext& context): context(context) {}
+    ~ParseErrorCatcher() noexcept(false) {
+      if (!unwindDetector.isUnwinding()) {
+        KJ_IF_MAYBE(e, exception) {
+          context.error(kj::str(
+              "*** ERROR CONVERTING PREVIOUS MESSAGE ***\n"
+              "The following error occurred while converting the message above.\n"
+              "This probably means the input data is invalid/corrupted.\n",
+              "Exception description: ", e->getDescription(), "\n"
+              "Code location: ", e->getFile(), ":", e->getLine(), "\n"
+              "*** END ERROR ***"));
+        }
+      }
+    }
+
+    void onRecoverableException(kj::Exception&& e) {
+      // Only capture the first exception, on the assumption that later exceptions are probably
+      // just cascading problems.
+      if (exception == nullptr) {
+        exception = kj::mv(e);
+      }
+    }
+
+  private:
+    kj::ProcessContext& context;
+    kj::Maybe<kj::Exception> exception;
+    kj::UnwindDetector unwindDetector;
+  };
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* Data, size_t Size)
+{
+    kj::TopLevelProcessContext context("fuzz");
+    ParseErrorCatcher parseErrorCatcher(context);
```

Comment:
You don't need `ParseErrorCatcher` since the tests below are wrapped in `runCatchingExceptions`. In fact, `ParseErrorCatcher` is probably detrimental since it actually causes exceptions to be ignored and execution to continue as if they didn't happen, which isn't how people would normally use the library. (It's used in the CLI tool because it's useful for debugging to decode as much as possible even if some parts of the message are corrupted.)

Reply:
If I remember correctly, not having `ParseErrorCatcher` resulted in having errors such as `expected n >= minBytes [2 >= 8]; Premature EOF`

I will check again

Reply:
Those errors are expected when the input is corrupt. `ParseErrorCatcher` is just causing catching the errors and discarding them. But again, we use `runCatchingExceptions` below for that same purpose.

Reply:
Ok, will try with only `runCatchingExceptions` and no `ParseErrorCatcher`

Reply:
Ok, I had to include `capnp::InputStreamMessageReader reader(ais);` in the `runCatchingExceptions` as it was the one throwing Premature EOF